python/mrc/_pymrc/src/segment.cpp

```diff
@@ -257,6 +258,68 @@ std::shared_ptr<mrc::segment::ObjectProperties> build_source(mrc::segment::IBuil
     return self.construct_object<PythonSource<PyHolder>>(name, wrapper);
 }
 
+class SubscriberFuncWrapper : public mrc::pymrc::PythonSource<PyHolder>
+{
+  public:
+    using base_t = mrc::pymrc::PythonSource<PyHolder>;
+    using typename base_t::source_type_t;
+    using typename base_t::subscriber_fn_t;
+
+    SubscriberFuncWrapper(py::function gen_factory) : PythonSource(build()), m_gen_factory{std::move(gen_factory)} {}
+
+    ~SubscriberFuncWrapper() override
+    {
+        {
+            AcquireGIL gil;
+            py::object kill = std::move(m_gen_factory);
+        }
+    }
+
+  private:
+    subscriber_fn_t build()
+    {
+        return [this](rxcpp::subscriber<source_type_t> subscriber) {
+            auto& ctx = runnable::Context::get_runtime_context();
+
+            try
+            {
+                DVLOG(10) << ctx.info() << " Starting source";
+                py::gil_scoped_acquire gil;
+                py::object py_sub    = py::cast(subscriber);
+                py::iterator py_iter = py::cast<py::iterator>(m_gen_factory(std::move(py_sub)));
+                PyIteratorWrapper iter_wrapper{std::move(py_iter)};
+
+                for (auto next_val : iter_wrapper)
+                {
+                    //  Only send if its subscribed. Very important to ensure the object has been moved!
+                    if (subscriber.is_subscribed())
+                    {
+                        py::gil_scoped_release no_gil;
+                        subscriber.on_next(std::move(next_val));
+                    }
+                    else
+                    {
+                        DVLOG(10) << ctx.info() << " Source unsubscribed. Stopping";
+                        break;
+                    }
+                }
+
+            } catch (const std::exception& e)
+            {
+                LOG(ERROR) << ctx.info() << "Error occurred in source. Error msg: " << e.what();
+
+                subscriber.on_error(std::current_exception());
+                return;
+            }
+            subscriber.on_completed();
+
+            DVLOG(10) << ctx.info() << " Source complete";
+        };
+    }
+
+    py::function m_gen_factory{};
```

Comment:
We should use the `PyFuncWrapper` here to ensure we have the GIL when this is destroyed (then you dont need the destructor).